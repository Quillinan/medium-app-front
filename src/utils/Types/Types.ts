export interface AuthContextType {
  setAuth: (authToken: string | null) => void;
  removeAuth: () => void;
  token: string | null;
}

export interface ErrorResponse {
  code: string;
  message: string;
  detailedMessage: string;
  helpUrl: string;
  details: null;
}

export interface Birthday {
  NOME: string;
  DTNASCIMENTO: number;
}

export interface TableBirthdayProps {
  data: { NOME: string; DTNASCIMENTO: number }[];
}

export interface Post {
  id: number;
  title: string;
  subtitle: string;
  content: string;
  createdAt: Date;
  updatedAt: Date;
  authorId: string;
  authorName: string;
}
