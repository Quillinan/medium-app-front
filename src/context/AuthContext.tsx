import { PublicClientApplication } from '@azure/msal-browser';
import { MsalProvider } from '@azure/msal-react';
import { AuthContextType } from '@utils/types';
import { createContext, ReactNode, useState } from 'react';

export const AuthContext = createContext<AuthContextType | undefined>(
  undefined
);

const msalInstance = new PublicClientApplication({
  auth: {
    clientId: '1efce811-7a11-4e35-98a8-d85f7b169637',
    authority: 'https://login.microsoftonline.com/gruposendas.onmicrosoft.com',
    redirectUri: 'http://localhost:5173/home',
  },
});

export const AuthProvider: React.FC<{ children: ReactNode }> = ({
  children,
}) => {
  const [token, setToken] = useState<string | null>(null);

  const setAuth = (authToken: string | null) => {
    if (authToken) {
      sessionStorage.setItem('isAuthenticated', 'true');
      sessionStorage.setItem('authToken', authToken);
    } else {
      sessionStorage.removeItem('isAuthenticated');
      sessionStorage.removeItem('authToken');
    }
    setToken(authToken);
  };

  const removeAuth = () => {
    sessionStorage.removeItem('isAuthenticated');
    sessionStorage.removeItem('authToken');
    setToken(null);
  };

  return (
    <AuthContext.Provider value={{ setAuth, removeAuth, token }}>
      <MsalProvider instance={msalInstance}>{children}</MsalProvider>
    </AuthContext.Provider>
  );
};
