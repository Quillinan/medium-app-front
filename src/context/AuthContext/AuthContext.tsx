import { PublicClientApplication } from '@azure/msal-browser';
import { MsalProvider } from '@azure/msal-react';
import { AuthContextType } from '@utils/types';
import { createContext, ReactNode, useState } from 'react';

export const AuthContext = createContext<AuthContextType | undefined>(
  undefined
);

const deploy = true;
let clientId = import.meta.env.VITE_API_CLIENTIDDEV;
let redirectUri = import.meta.env.VITE_API_REDURLDEV;

if (deploy) {
  clientId = import.meta.env.VITE_API_CLIENTIDPROD;
  redirectUri = import.meta.env.VITE_API_REDURLPROD;
}

const msalInstance = new PublicClientApplication({
  auth: {
    clientId: clientId,
    authority: 'https://login.microsoftonline.com/gruposendas.onmicrosoft.com',
    redirectUri: redirectUri,
  },
});

export const AuthProvider: React.FC<{ children: ReactNode }> = ({
  children,
}) => {
  const [token, setToken] = useState<string | null>(null);

  const setAuth = (authToken: string | null) => {
    if (authToken) {
      sessionStorage.setItem('isAuthenticated', 'true');
      sessionStorage.setItem('authToken', authToken);
    } else {
      sessionStorage.removeItem('isAuthenticated');
      sessionStorage.removeItem('authToken');
    }
    setToken(authToken);
  };

  const removeAuth = () => {
    sessionStorage.removeItem('isAuthenticated');
    sessionStorage.removeItem('authToken');
    setToken(null);
  };

  return (
    <AuthContext.Provider value={{ setAuth, removeAuth, token }}>
      <MsalProvider instance={msalInstance}>{children}</MsalProvider>
    </AuthContext.Provider>
  );
};
