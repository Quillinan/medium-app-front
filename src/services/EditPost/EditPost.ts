import handleError from '@errors/HandleError/HandleError';
import { apiUrl, putApi } from '@services/Api/Api';
import { Post, ErrorResponse, EditPostData } from '@utils/Types/Types';

export const editPost = async (
  postData: EditPostData,
  postId: string
): Promise<Post | ErrorResponse | undefined> => {
  const url = `${apiUrl}api/Post/${postId}`;
  const formData = new FormData();

  if (postData.title) {
    formData.append('Title', postData.title);
  }
  if (postData.subtitle) {
    formData.append('Subtitle', postData.subtitle);
  }
  if (postData.content) {
    formData.append('Content', postData.content);
  }
  if (postData.coverImage) {
    formData.append('CoverImageData', postData.coverImage);
  }

  try {
    const response = await putApi(url, formData);
    return response as Post;
  } catch (error: any) {
    return handleError(error);
  }
};
