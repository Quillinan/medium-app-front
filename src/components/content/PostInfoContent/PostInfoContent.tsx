import React, { useState, useContext } from 'react';
import { Post, EditPostData } from '@utils/Types/Types';
import { AuthContext } from '@context/AuthContext/AuthContext';
import CreatePostContent from '../CreatePostContent/CreatePostContent';
import EditDeleteButtons from '@components/buttons/EditDeleteButtons/EditDeleteButtons';
import { editPost } from '@services/EditPost/EditPost';

interface PostInfoContentProps {
  post: Post;
}

const PostInfoContent: React.FC<PostInfoContentProps> = ({ post }) => {
  const authContext = useContext(AuthContext);
  const userId = authContext?.uniqueId;
  const [isEditing, setIsEditing] = useState(false);
  const [postData, setPostData] = useState<EditPostData>({
    title: post.title,
    subtitle: post.subtitle,
    content: post.content,
    coverImage: null,
  });

  const handleSave = async () => {
    console.log('Salvei!');
    try {
      const response = await editPost(postData, String(post.id));
      if (response) {
        console.log('Post atualizado com sucesso:', response);
        setIsEditing(false);
      }
    } catch (error) {
      console.error('Erro ao atualizar o post:', error);
    }
  };

  return (
    <div className='flex flex-col items-center w-full'>
      {userId === post.authorId && (
        <EditDeleteButtons
          isEditing={isEditing}
          onToggleEdit={() => setIsEditing(!isEditing)}
          onSave={handleSave}
          postId={String(post.id)}
        />
      )}

      {isEditing ? (
        <CreatePostContent
          setCoverImage={image =>
            setPostData(prevData => ({ ...prevData, coverImage: image }))
          }
          coverImage={postData.coverImage || null}
          initialPostData={{
            title: postData.title || '',
            subtitle: postData.subtitle || '',
            content: postData.content || '',
            coverImageUrl: post.coverImageUrl,
          }}
          isEditing={true}
        />
      ) : (
        <div className='flex flex-col h-full p-4 space-y-4 w-1/3 mt-6'>
          <h1 className='font-bold mb-3' style={{ fontSize: '42px' }}>
            {post.title}
          </h1>
          <h2 className='text-gray-700 mb-2' style={{ fontSize: '22px' }}>
            {post.subtitle}
          </h2>
          <p className='mt-4 text-gray-500' style={{ fontSize: '16px' }}>
            Autor: {post.authorName} | Criado em:{' '}
            {new Date(post.createdAt).toLocaleDateString('pt-BR')}
          </p>
          {post.coverImageUrl && (
            <div className='flex justify-center mt-4'>
              <img
                src={post.coverImageUrl}
                alt={post.title}
                className='w-full h-auto m-5'
              />
            </div>
          )}
          <div
            className='mt-4 w-full'
            style={{ fontSize: '20px' }}
            dangerouslySetInnerHTML={{ __html: post.content }}
          />
        </div>
      )}
    </div>
  );
};

export default PostInfoContent;
