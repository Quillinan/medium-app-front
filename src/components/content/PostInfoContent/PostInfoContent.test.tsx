import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';
import PostInfoContent from './PostInfoContent';
import { editPost } from '@services/EditPost/EditPost';
import { AuthContext } from '@context/AuthContext/AuthContext';
import { Post } from '@utils/Types/Types';
import { Mock, vi } from 'vitest';

vi.mock('@services/EditPost/EditPost', () => ({
  editPost: vi.fn(),
}));

const mockPost: Post = {
  id: 1,
  title: 'Test Title',
  subtitle: 'Test Subtitle',
  content: '<p>Test Content</p>',
  createdAt: new Date(),
  updatedAt: new Date(),
  authorId: '12345',
  authorName: 'Test Author',
  coverImageUrl: '',
};

const mockAuthContext = {
  setAuth: vi.fn(),
  removeAuth: vi.fn(),
  token: 'mockToken',
  uniqueId: '12345',
  name: 'Mock User',
};

describe('PostInfoContent Component', () => {
  afterEach(() => {
    vi.clearAllMocks();
  });

  beforeAll(() => {
    vi.spyOn(console, 'error').mockImplementation(() => {});
  });

  afterAll(() => {
    (console.error as Mock).mockRestore();
  });

  it('displays error message if editPost fails', async () => {
    const mockError = new Error('Network Error');
    (editPost as Mock).mockRejectedValueOnce(mockError);

    render(
      <AuthContext.Provider value={mockAuthContext}>
        <PostInfoContent post={mockPost} />
      </AuthContext.Provider>
    );

    const editButton = screen.getByRole('button', { name: /Editar/i });
    fireEvent.click(editButton);

    const saveButton = screen.getByRole('button', { name: /Salvar/i });
    fireEvent.click(saveButton);

    await waitFor(() => {
      expect(
        screen.getByText(/Erro ao atualizar o post:/i)
      ).toBeInTheDocument();
    });
  });

  it('renders post content in view mode', () => {
    render(
      <AuthContext.Provider value={mockAuthContext}>
        <PostInfoContent post={mockPost} />
      </AuthContext.Provider>
    );

    expect(screen.getByText(mockPost.title)).toBeInTheDocument();
    expect(screen.getByText(mockPost.subtitle)).toBeInTheDocument();

    expect(
      screen.getByText(
        new RegExp(`Autor: ${mockPost.authorName} \\| Criado em:.*`)
      )
    ).toBeInTheDocument();
  });

  it('allows the author to switch to edit mode', () => {
    render(
      <AuthContext.Provider value={mockAuthContext}>
        <PostInfoContent post={mockPost} />
      </AuthContext.Provider>
    );

    const editButton = screen.getByRole('button', { name: /Editar/i });
    fireEvent.click(editButton);

    expect(screen.getByDisplayValue(mockPost.title)).toBeInTheDocument();
    expect(screen.getByDisplayValue(mockPost.subtitle)).toBeInTheDocument();
  });

  it('calls editPost and saves changes when the save button is clicked', async () => {
    const mockUpdatedTitle = 'Updated Title';
    (editPost as jest.Mock).mockResolvedValue({
      ...mockPost,
      title: mockUpdatedTitle,
    });

    render(
      <AuthContext.Provider value={mockAuthContext}>
        <PostInfoContent post={mockPost} />
      </AuthContext.Provider>
    );

    const editButton = screen.getByRole('button', { name: /Editar/i });
    fireEvent.click(editButton);

    const titleInput = screen.getByDisplayValue(mockPost.title);
    fireEvent.change(titleInput, { target: { value: mockUpdatedTitle } });

    const saveButton = screen.getByRole('button', { name: /Salvar/i });
    fireEvent.click(saveButton);

    await waitFor(() => {
      expect(editPost).toHaveBeenCalledWith(
        expect.objectContaining({ title: mockUpdatedTitle }),
        String(mockPost.id)
      );
    });
  });

  it('should match snapshot', () => {
    const { asFragment } = render(
      <AuthContext.Provider value={mockAuthContext}>
        <PostInfoContent post={mockPost} />
      </AuthContext.Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
