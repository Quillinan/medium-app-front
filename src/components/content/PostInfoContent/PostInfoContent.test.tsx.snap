// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PostInfoContent Component > should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col items-center w-full"
  >
    <div
      class="w-1/2 flex justify-end space-x-4 mt-6"
    >
      <button
        class="bg-yellow-500 text-white rounded px-4 py-1"
      >
        Editar
      </button>
      <button
        class="bg-red-500 text-white rounded px-4 py-1"
      >
        Apagar
      </button>
    </div>
    <div
      class="flex flex-col h-full p-4 space-y-4 w-1/3 mt-6"
    >
      <h1
        class="font-bold mb-3"
        style="font-size: 42px;"
      >
        Test Title
      </h1>
      <h2
        class="text-gray-700 mb-2"
        style="font-size: 22px;"
      >
        Test Subtitle
      </h2>
      <p
        class="mt-4 text-gray-500"
        style="font-size: 16px;"
      >
        Autor: Test Author | Criado em: 08/11/2024
      </p>
      <div
        class="mt-4 w-full"
        style="font-size: 20px;"
      >
        <p>
          Test Content
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
